src/Compilers/CSharp/Portable/Syntax/NonRecursiveRewriter.cs

```diff
@@ -0,0 +1,287 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph and
+    /// may replace or remove visited SyntaxNodes in depth-first order.
+    /// </summary>
+    public partial class NonRecursiveRewriter : CSharpSyntaxVisitor<NonRecursiveRewriter.Chunk>
+    {
+        public class Chunk
+        {
+            private int currentIndex;
+
+            internal Chunk(Func<SyntaxNodeOrToken> transform, SyntaxNodeOrToken[] children)
+            {
+                this.Transform = transform;
+                this.ChildNodes = children;
+                this.currentIndex = children.Length;
+            }
+
+            public Func<SyntaxNodeOrToken> Transform { get; private set; }
+
+            public SyntaxNodeOrToken[] ChildNodes { get; private set; }
+
+            public bool MoveNextChildNode()
+            {
+                return this.currentIndex-- > 0;
+            }
+
+            public SyntaxNodeOrToken GetCurrentChild()
+            {
+                return ChildNodes[this.currentIndex];
+            }
+        }
+
+        private Stack<SyntaxNodeOrToken> rewritenStack = new Stack<SyntaxNodeOrToken>();
+        protected internal Stack<SyntaxNodeOrToken> RewritenStack
+        {
+            get { return this.rewritenStack; }
+        }
+
+        public new SyntaxNode Visit(SyntaxNode node)
+        {
+            return this.Visit((SyntaxNodeOrToken)node);
+        }
+
+        public SyntaxNode Visit(SyntaxToken token)
+        {
+            return this.Visit((SyntaxNodeOrToken)token);
+        }
+
+        private SyntaxNode Visit(SyntaxNodeOrToken node)
+        {
+            if (node == null)
```

Comment:
can this ever be true?  SyntaxNodeOrToken is a struct.  So it should not be possible for it to be null.